### PR TITLE
ingest: date_index_name processor template resolution

### PR DESCRIPTION
**Pull request by jakelandis**
This change adds support for template snippet (e.g. `{{foo}}`) resolution
in the date_index_name processor. The following configuration options
will now resolve a templated value if so configured:

* index_name_prefix  (e.g "index_name_prefix": "myindex-`{{foo}}`-")
* date_rounding (e.g. "date_rounding" : "`{{bar}}`")
* index_name_format (e.g."index_name_format": "`{{baz}}`")

Example:
```
POST _ingest/pipeline/_simulate
{
  "pipeline" : {
    "processors" : [
      { "date_index_name": 
        { "field": "@timestamp", 
          "index_name_prefix": "myindex-{{event_type}}-", 
          "date_rounding" : "{{index-period}}",
          "index_name_format": "{{index_format}}"
        }
      }
    ]
  },
  "docs" : [
     {"_source": {
       "@timestamp" : "2018-07-04T12:00:00.000Z",
       "event_type" : "rare",
       "index-period" : "y",
       "index_format": "yyyy"
     }}
    ]
}
```

results in
```
{
  "docs": [
    {
      "doc": {
        "_index": "<myindex-rare-{2018||/y{yyyy|UTC}}>",
        "_type": "_type",
        "_id": "_id",
        "_source": {
          "@timestamp": "2018-07-04T12:00:00.000Z",
          "event_type": "rare",
          "index-period": "y",
          "index_format": "yyyy"
        },
        "_ingest": {
          "timestamp": "2018-07-10T18:01:22.750869Z"
        }
      }
    }
  ]
}
```

_edit_ updated example
_edit2_ updated example
_edit3_ removed note that is no longer relevant

### CODE REVIEW

File: modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/DateIndexNameProcessor.java [hunk 1]
```diff
@@ -94,7 +98,7 @@ public void execute(IngestDocument ingestDocument) throws Exception {
                     .append('}')
                 .append('>');
         String dynamicIndexName  = builder.toString();
-        ingestDocument.setFieldValue(IngestDocument.MetaData.INDEX.getFieldName(), dynamicIndexName);
+        ingestDocument.setFieldValue(IngestDocument.MetaData.INDEX.getFieldName(), ValueSource.wrap(dynamicIndexName, scriptService));
```
Comment (martijnvg) [inline modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/DateIndexNameProcessor.java]: I think it is better that the `indexNamePrefix`, `dateRounding` and `indexNameFormat` become `TemplateScript.Factory` typed fields, like in other processor implementations. ValueSource is more meant to template generic values and in this case we know the templated values are always strings. Also then it is not needed to pass down `ScriptService` into the processor implementation and then the change in `IngestDocument` is not needed too.

Comment (jakelandis) [reply modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/DateIndexNameProcessor.java]: @martijnvg - thanks for the review !  the changes requested are at https://github.com/elastic/elasticsearch/pull/31841/commits/f9a6ec1e18c054b2184f03b6cfd429a940d75f1c.  This PR should be ready for re-review.

**Comment by jakelandis** [comment]
Jenkins test this please

**Comment by jakelandis** [comment]
Jenkins test this please

**Review by martijnvg — APPROVED** [review]
LGTM 👍

**Comment by jakelandis** [comment]
@martijnvg - thanks!  merged and will backport to `6.x`.

**Pull request merged**